sympy/core/numbers.py

```diff
@@ -2641,6 +2748,119 @@ def _eval_simplify(self, **kwargs):
                     return AlgebraicNumber(r)
         return self
 
+    @property
+    def is_primitive_elt(self):
+        r"""
+        Say whether this algebraic number $\alpha \in \mathbb{Q}(\theta)$ is
+        equal to the primitive element $\theta$ for its field.
+        """
+        return self.coeffs() == [1, 0]
+
+    def primitive_elt(self):
+        r"""
+        Get the primitive element $\theta$ for the number field
+        $\mathbb{Q}(\theta)$ to which this algebraic number $\alpha$ belongs.
+
+        Returns
+        =======
+
+        AlgebraicNumber
+
+        """
+        if self.is_primitive_elt:
+            return self
+        return AlgebraicNumber(self, coeffs=[1, 0])
+
+    def to_primitive_elt(self, prec=15):
+        r"""
+        Convert ``self`` to an :py:class:`~.AlgebraicNumber` instance that is
+        equal to its own primitive element.
+
+        Explanation
+        ===========
+
+        Since an :py:class:`~.AlgebraicNumber` stores both the minimal
+        polynomial, and a particular root value, for its primitive element,
+        it is sometimes more convenient to work with instances that equal their
+        own primitive element.
+
+        Examples
+        ========
+
+        >>> from sympy import sqrt, to_number_field
+        >>> from sympy.abc import x
+        >>> a = to_number_field(sqrt(2), sqrt(2) + sqrt(3))
+
+        The :py:class:`~.AlgebraicNumber` ``a`` represents the number
+        $\sqrt{2}$ in the field $\mathbb{Q}(\sqrt{2} + \sqrt{3})$. Rendering
+        ``a`` as a polynomial,
+
+        >>> a.as_poly().as_expr(x)
+        x**3/2 - 9*x/2
+
+        reflects the fact that $\sqrt{2} = \theta^3/2 - 9 \theta/2$, where
+        $\theta = \sqrt{2} + \sqrt{3}$.
+
+        ``a`` is not equal to its own primitive element. Its minpoly
+
+        >>> a.minpoly.as_poly().as_expr(x)
+        x**4 - 10*x**2 + 1
+
+        is that of $\theta$.
+
+        Converting to a primitive element,
+
+        >>> a_prim = a.to_primitive_elt()
+        >>> a_prim.minpoly.as_poly().as_expr(x)
+        x**2 - 2
+
+        we obtain an :py:class:`~.AlgebraicNumber` whose ``minpoly`` is that of
+        the number itself.
+
+        Parameters
+        ==========
+
+        prec : int, optional (default=15)
+            Decimal places of precision for determining to which root of its
+            minimal polynomial this number is equal.
+
+        Returns
+        =======
+
+        AlgebraicNumber
+
+        See Also
+        ========
+
+        is_primitive_elt
+
+        """
+        if self.is_primitive_elt:
+            return self
+        m = self.minpoly_of_elt()
+        r0 = self.n(prec)
+        closest_root, min_distance = None, oo
+        for r in m.all_roots():
+            d = abs(r0 - r.n(prec))
+            if d < min_distance:
+                min_distance = d
+                closest_root = r
+        return AlgebraicNumber((m, closest_root))
```

Comment:
@oscarbenjamin Do you know if there's an existing function for this problem? I couldn't find one, so wrote an ad hoc solution.

You have a `Poly` `m`, and you have a number (in this case `self`), and you know that the number is equal to one of the roots of `m`. You want to determine which of the roots it is, and then you want to obtain an exact expression for that root (such as a `CRootOf`, or an explicit expression using radicals etc.).

Reply:
I'm not sure there is a function but one could be added to rootisolation.py. The key function is `dup_isolate_all_roots` which gives bounding intervals for each of the roots. Then evalf (i.e. .n) can tell us if our expression is in the bounding interval.

Reply:
This functions is the only thing I'm not sure about

Reply:
Maybe this could be based on the same idea that is used in `_choose_factor`:
https://github.com/sympy/sympy/blob/551c16facb2081c7b7a085aeaf20316244a1da01/sympy/polys/numberfields/minpoly.py#L80-L86
All distances are computed and two least ones are compared. It they have different orders of magnitude (probably even less than `10**6` would suffice), then the minimum is chosen, otherwise the precision in incremented in a loop. There is no need to give a precision parameter as an argument.

Reply:
Wait...... First, thanks @oscarbenjamin and @jksuom for your ideas. I think you're on the right track for the problem
I asked to solve. However, maybe it's not needed here.

What we're talking about could go in a method called `to_CRootOf()`, which would turn an `AlgebraicNumber` into a `CRootOf` equal to the same complex number.

But I think all we need here is any `Expr` that can be `evalf`-ed to arbitrary precision, and `self` already is that.
So the whole method becomes:

```python
def to_primitive_elt(self):
    if self.is_primitive_elt:
        return self
    m = self.minpoly_of_elt()
    return AlgebraicNumber((m, self))
```

Sorry for the confusion! I didn't see this earlier.

Reply:
Cancel that. This turns out to be quite tricky.

## TL;DR

I think we were on the right track in the first place, and I'll push a new commit that does a better job of determining the root.

## Details

It seems that using one `AlgebraicNumber` as the `root` arg to another `AlgebraicNumber` is going to be very error prone and should probably be avoided.

As a demonstration of what can happen, let's start with an `AlgebraicNumber` `a` and its minpoly `m`:

```python
>>> from sympy import sqrt, to_number_field, AlgebraicNumber
>>> a = to_number_field(sqrt(2), sqrt(2) + sqrt(3))
>>> a.as_expr()
sqrt(2)
>>> m = a.minpoly_of_elt()
```

and use these as the `(m, r)` when constructing a new `AlgebraicNumber` `b`:

```python
>>> b = AlgebraicNumber((m, a))
>>> b.as_expr()
sqrt(2)
```

So far, so good.

But now, due to the funny way `AlgebraicNumber.__new__()` treats its first arg, when that arg is an `AlgebraicNumber`,

https://github.com/sympy/sympy/blob/519f1b75eda8830189b1385197f2c4fb49d333aa/sympy/core/numbers.py#L2541-L2542

`b` is not going to be picklable:

```python
>>> import pickle
>>> c = pickle.loads(pickle.dumps(b))
>>> c.as_expr()
sqrt(2) + sqrt(3)
```

Yipes!

But maybe we can repair pickling by changing this line:

https://github.com/sympy/sympy/blob/519f1b75eda8830189b1385197f2c4fb49d333aa/sympy/core/numbers.py#L2564

to

```python
        sargs = (expr, scoeffs)
```

which seems totally reasonable. Let our args be exactly as they were passed to us.

But this turns out to have too many far-flung ramifications.

For just one example, [this unit test](https://github.com/sympy/sympy/blob/519f1b75eda8830189b1385197f2c4fb49d333aa/sympy/polys/numberfields/tests/test_minpoly.py#L416-L422) is broken by it, because this method:

https://github.com/sympy/sympy/blob/519f1b75eda8830189b1385197f2c4fb49d333aa/sympy/polys/domains/gaussiandomains.py#L320-L323

expects the first arg of an `AlgebraicNumber` to always equal its root. In how many other places might similar expectations be embedded? This approach begins to look like much more trouble than it's worth, so I'll revert to the original approach.

Reply:
I have to say that I've never quite understood how `AlgebraicNumber` works. Maybe it is not used consistently and should be fixed.

Reply:
It seems that an instance of `AlgebraicNumber` is practically always the same as the primitive element except in some tests. Maybe `coeffs` could be deprecated (and the tests removed).

Reply:
The possibility did cross my mind, but I think there are strong reasons for keeping it.

* A computer algebra system should have a class that represents "an element of a number field".
Such elements are usually thought of as polynomials in the primitive element, and the class should
provide things like printing support for displaying the element in this form. `AlgebraicNumber`
currently provides all this, but it will cease to if we give up `coeffs`.

* The `ANP` class does things the same way -- i.e. it represents the minimal polynomial of the
primitive element, plus the coeffs of the given element -- but this is on the `polys` side of things.
It is not an `Expr`.

* An `AlgebraicNumber` with `coeffs` is needed as the return value of the public `to_number_field()`
function.

* If we give up `coeffs`, then we just have a `minpoly` and a `root`, in which case we're essentially
the same thing as a `CRootOf`. (Which, to be honest, is a fine representation of an algebraic number for
certain applications -- but then it would be redundant.)

I guess those are the main reasons that I have considered.

Also, I'm still hoping to implement more things from Cohen. So, you're probably right that right now the library mostly treats these things as their primitive elements, but I'm hoping to add some cool stuff which will make an "element of a number field" class of this kind more important!

Reply:
I think that `AlgebraicNumber` should just be fixed to work properly.

Reply:
> But now, due to the funny way AlgebraicNumber.__new__() treats its first arg, when that arg is an AlgebraicNumber,

There `expr.coeffs` should be checked as well. That should be combined with the newly provided coeffs using e.g. `dup_compose`

Reply:
But isn't it a backwards compatibility issue? Otherwise I'd be all for changing the behavior here. (I think the special case should just be eliminated.)

Reply:
Can you give an example of what would be incompatible if this was just fixed?

Reply:
I guess the real question I was raising was where we draw the line between breaking change and bugfix. But it dawns on me now that this entire special case of `AlgebraicNumber.__new__(expr, coeffs, ...)` where `expr` is an `AlgebraicNumber` could be regarded as a part of the same bug this PR sets out to fix; namely, places in the code where the `coeffs` of an `AlgebraicNumber` were forgotten/ignored.

So, I think the composing of polynomials here is clever, it brings the coeffs back into play, and it means that on unpickling we recover a number which is the same as a complex number, all of which is good. But unfortunately the internal structure still may differ. Example:

Adding this in `.__new__()`:

```python
        if expr.is_AlgebraicNumber:
            from sympy.polys.densetools import dup_compose
            c = dup_compose(rep.rep, expr.rep.rep, dom)
            rep = DMP.from_list(c, 0, dom)
```

we now get:

```python
>>> import pickle
>>> a = to_number_field(sqrt(2), sqrt(2) + sqrt(3))
>>> m = a.minpoly_of_elt()
>>> b = AlgebraicNumber((m, a))
>>> c = pickle.loads(pickle.dumps(b))
>>> for e in [b, c]:
>>>     print('--------')
>>>     print('As a complex number:         ', e.as_expr())
>>>     print('Our coeffs:                  ', e.rep)
>>>     print('Minpoly of primitive element:', e.minpoly)
>>>     print('Primitive element:           ', e.root)
>>> print('--------')
>>> print('Equal after pickling:', b == c)
--------
As a complex number:          sqrt(2)
Our coeffs:                   DMP([mpq(1,1), mpq(0,1)], QQ, None)
Minpoly of primitive element: PurePoly(_x**2 - 2, _x, domain='QQ')
Primitive element:            sqrt(2)
--------
As a complex number:          sqrt(2)
Our coeffs:                   DMP([mpq(1,2), mpq(0,1), mpq(-9,2), mpq(0,1)], QQ, None)
Minpoly of primitive element: PurePoly(_x**4 - 10*_x**2 + 1, _x, domain='QQ')
Primitive element:            sqrt(2) + sqrt(3)
--------
Equal after pickling: False
```

I'm thinking the fix could instead be as simple as just deleting the special case,

https://github.com/sympy/sympy/blob/52f606a503cea5e9588de14150ccb9f7f9ed4752/sympy/core/numbers.py#L2541-L2542

altogether. Then we do test as equal after pickling:

```python
--------
As a complex number:          -9*sqrt(2) + sqrt(3)/2 + sqrt(2) + sqrt(3)**3/2
Our coeffs:                   DMP([mpq(1,1), mpq(0,1)], QQ, None)
Minpoly of primitive element: PurePoly(_x**2 - 2, _x, domain='QQ')
Primitive element:            -9*sqrt(2) + sqrt(3)/2 + sqrt(2) + sqrt(3)**3/2
--------
As a complex number:          -9*sqrt(2) + sqrt(3)/2 + sqrt(2) + sqrt(3)**3/2
Our coeffs:                   DMP([mpq(1,1), mpq(0,1)], QQ, None)
Minpoly of primitive element: PurePoly(_x**2 - 2, _x, domain='QQ')
Primitive element:            -9*sqrt(2) + sqrt(3)/2 + sqrt(2) + sqrt(3)**3/2
--------
Equal after pickling: True

```

Note that the printing under `.as_expr()` is also changed in this case. This is because we are now succeeding in restoring `self.root` as an instance of `AlgebraicNumber`, which stays atomic in simplifications.

I think this is a good thing. If the old behavior is desired, then `b` can just be constructed a little differently:

```python
>>> b = AlgebraicNumber((m, a.to_root()))
>>> print(b.as_expr())
sqrt(2)
```

Here I'm using the `to_root()` method which I've now implemented, and have just pushed. In this case `a` is transformed into a `Pow`, which does participate in expression simplifications.

It's complex. I want to continue thinking about it a bit more...

Reply:
Okay, I see the problem now. Your examples use lots of complicated methods but it's actually just this:
```python
In [15]: a = AlgebraicNumber(sqrt(2), [1, 2])

In [16]: a
Out[16]: √2 + 2

In [17]: AlgebraicNumber(a)
Out[17]: √2
```
That's a clear bug to be fixed. Does the fix need to be much more complicated than this:
```diff
diff --git a/sympy/core/numbers.py b/sympy/core/numbers.py
index e87b039..93c07c8 100644
--- a/sympy/core/numbers.py
+++ b/sympy/core/numbers.py
@@ -2536,7 +2536,7 @@ def __new__(cls, expr, coeffs=None, alias=None, **args):
                 from sympy.polys.polytools import Poly
                 minpoly = Poly(minpoly)
         elif expr.is_AlgebraicNumber:
-            minpoly, root = expr.minpoly, expr.root
+            minpoly, root, coeffs = expr.minpoly, expr.root, expr.coeffs()
         else:
             minpoly, root = minimal_polynomial(
                 expr, args.get('gen'), polys=True), expr
```
Could also check if coeffs was provided as well as an AlgebraicNumber and if so use `dup_compose`.

Reply:
Okay, this has the virtue of sticking closer to existing behavior, so I think it's good.

But while we're making `AlgebraicNumber(a)` behave more like a "copy constructor" when `a` is an `AlgebraicNumber`, we should probably also copy the `self.alias` from `a`, allowing this to be overridden by any `alias` that may have been passed.

Also, we need to modify the case where `expr` is an ordered pair `(m, r)`. If `r` is an `AlgebraicNumber` we have to convert it with `r = r.to_root()`. This (a) repairs pickling, and (b) makes this case consistent with the others.

I've pushed some code that implements this, and adapts the docs.

Reply:
Scratch the bit about `r = r.to_root()`. I tried it, but reverted it since it broke some things.

Anyway this was intended to address a separate issue -- pickling of `AlgebraicNumber` and `AlgebraicField` -- and I'd like to handle this in a separate PR.